clash-cores/src/Clash/Cores/Sgmii/PcsReceive.hs

```diff
@@ -0,0 +1,520 @@
+{-# LANGUAGE RecordWildCards #-}
+
+module Clash.Cores.Sgmii.PcsReceive where
+
+import Clash.Cores.Sgmii.Common
+import Clash.Cores.Sgmii.EbTb
+import Clash.Prelude
+
+-- | Defines all possible valid 'checkEnd' results, as well as @None@ for no
+--   valid result
+data CheckEnd
+  = K285DK285
+  | K285D022D000
+  | K285D215D000
+  | RRK285
+  | RRR
+  | RRS
+  | TRK285
+  | TRR
+  | None
+  deriving (Eq, Show)
+
+-- | State type of 'pcsReceive'. This contains all states as they are defined in
+--   IEEE 802.3 Clause 36, with the addition of two initialization states that
+--   are used to fill the @cw@ input shift register and the exeception of the
+--   states @CARRIER_DETECT@, @RECEIVE@ and @EPD2_CHECK_END@ as these do not act
+--   upon a received code group. The transitions of these states are embedded in
+--   the states that usually transition to either of these states.
+data PcsReceiveState
+  = Init1
+  | Init2
+  | WaitForK {_receiving :: Bool}
+  | RxK {_receiving :: Bool}
+  | RxCB {_receiving :: Bool}
+  | RxCC {_receiving :: Bool, _hist :: DataWord}
+  | RxCD {_receiving :: Bool, _hist :: DataWord, _rxConfReg :: ConfReg}
+  | RxInvalid {_receiving :: Bool}
+  | IdleD {_receiving :: Bool}
+  | FalseCarrier {_receiving :: Bool}
+  | StartOfPacket {_receiving :: Bool}
+  | EarlyEnd {_receiving :: Bool}
+  | TriRri {_receiving :: Bool}
+  | TrrExtend {_receiving :: Bool}
+  | PacketBurstRrs {_receiving :: Bool}
+  | ExtendErr {_receiving :: Bool}
+  | EarlyEndExt {_receiving :: Bool}
+  | RxData {_receiving :: Bool, _hist :: DataWord}
+  | RxDataError {_receiving :: Bool, _hist :: DataWord}
+  | LinkFailed {_receiving :: Bool}
+  deriving (Generic, NFDataX, Eq, Show)
+
+-- | Calculate the bit difference between two code groups
+bitDifference ::
+  -- | First code group
+  BitVector 10 ->
+  -- | Second code group
+  BitVector 10 ->
+  -- | Bit difference
+  Index 11
+bitDifference cg0 cg1 = foldl f 0 $ map bitCoerce $ bv2v $ xor cg0 cg1
+ where
+  f a b = if b then a + 1 else a
```

Comment:
I've got no clue what this does.
Please elaborate a bit further or add a doctest that shows the behavior

Reply:
I hope the new comment makes it more clear. What does "add a doctest" mean?